docs/developers/developer-reference/smart-contract-developer-best-practices.md

```diff
@@ -266,44 +266,68 @@ The first approach looks a lot simpler, just a `ManagedVec` of tuples. But, the
+
+The main issue is that basic Rust types are quite eager with dynamic memory allocation: they ask for more memory than they actually need. For ordinary programs, this is great for performance, but for smart contracts, where every instruction costs gas, can be quite impactful, on both cost and even runtime failures.
+
+The alternative is to use **managed types** instead of the usual Rust types. All managed types, like `BigUint`, `TokenIdentifier`, `ManagedBuffer` etc. store all their contents inside the VM's memory, as opposed to the contract memory, so they have a great performance advantage. But you don't need to be concerned with "where" the contents are, because managed types automatically keep track of the contents with help from the VM.
+
+The managed types work by only storing a `handle` within the contract memory, which is a `u32` index, while the actual payload resides in reserved VM memory. So whenever you have to add two `BigUint`s for example, the `+` operation in your code will only pass the three handles: the result, the first operand, and the second operand. This way, there is very little data being passed around, which in turn makes everything cheaper. And since these types only store a handle, their memory allocation is fixed in size, so it can be allocated on the stack instead of having to be allocated on the heap.
+
+:::warning
+If you need to update older code to take advantage of managed types, please take the time to understand the changes you need to make. Such an update is important and cannot be done automatically.
+:::
 
 ### Base Rust types vs managed types
 
-Below is a table of unmanaged types and their managed counterparts:  
+Below is a table of unmanaged types (basic Rust types) and their managed counterparts, provided by the Elrond framework:
 
 | Unmanaged | Managed |
 | :---: | :---: |
-| &[u8] | TokenIdentifier |
-| - | BigUint |
-| BoxedBytes | ManagedBuffer |
-| String | ManagedBuffer |
-| VarArgs | ManagedVarArgs |
-| Vec | ManagedVec |
-| MultiResultVec | ManagedMultiResultVec |
+| `&[u8]` | `TokenIdentifier` |
+| - | `BigUint` |
+| `BoxedBytes` | `ManagedBuffer` |
+| `String` | `ManagedBuffer` |
+| `VarArgs` | `ManagedVarArgs` |
+| `Vec` | `ManagedVec` |
+| `MultiResultVec` | `ManagedMultiResultVec` |
+
+In most cases, the managed types can be used as drop-in replacements for the basic Rust types. For a simple example, see [BigUint Operations](#biguint-operations).
+
+:::warning
+Make sure you migrate to the managed types **incrementally** and **thoroughly test your code** before even considering deploying to the mainnet.
+:::
+
+:::tip
+You can use the `erdpy contract report` command to verify whether your contract still requires dynamic allocation or not.
+:::
 
```

Comment:
Let's also add a section about ArrayVec, which is easy to use, safe & fully integrated in elrond-wasm.
Only downside is that you need a max cap.